invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalConsumableUsageInArrear.java

```diff
@@ -0,0 +1,245 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import org.joda.time.LocalDate;
+import org.killbill.billing.callcontext.InternalTenantContext;
+import org.killbill.billing.catalog.api.CatalogApiException;
+import org.killbill.billing.catalog.api.TieredBlock;
+import org.killbill.billing.catalog.api.Usage;
+import org.killbill.billing.invoice.api.InvoiceItem;
+import org.killbill.billing.invoice.model.UsageInvoiceItem;
+import org.killbill.billing.invoice.usage.details.UsageConsumableInArrearDetail;
+import org.killbill.billing.invoice.usage.details.UsageConsumableInArrearTierUnitDetail;
+import org.killbill.billing.invoice.usage.details.UsageInArrearDetail;
+import org.killbill.billing.usage.RawUsage;
+import org.killbill.billing.usage.api.RolledUpUnit;
+import org.killbill.billing.util.config.definition.InvoiceConfig.UsageDetailMode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+
+import static org.killbill.billing.invoice.usage.UsageUtils.getConsumableInArrearTieredBlocks;
+
+public class ContiguousIntervalConsumableUsageInArrear extends ContiguousIntervalUsageInArrear {
+
+    private static final Logger log = LoggerFactory.getLogger(ContiguousIntervalConsumableUsageInArrear.class);
+
+    public ContiguousIntervalConsumableUsageInArrear(final Usage usage,
+                                                     final UUID accountId,
+                                                     final UUID invoiceId,
+                                                     final List<RawUsage> rawSubscriptionUsage,
+                                                     final LocalDate targetDate,
+                                                     final LocalDate rawUsageStartDate,
+                                                     final UsageDetailMode usageDetailMode,
+                                                     final InternalTenantContext internalTenantContext) {
+        super(usage, accountId, invoiceId, rawSubscriptionUsage, targetDate, rawUsageStartDate, usageDetailMode, internalTenantContext);
+    }
+
+    @Override
+    protected void populateResults(final LocalDate startDate, final LocalDate endDate, final Iterable<InvoiceItem> billedItems, final BigDecimal billedUsage, final BigDecimal toBeBilledUsage, final UsageInArrearDetail toBeBilledUsageDetails, final boolean areAllBilledItemsWithDetails, final List<InvoiceItem> result) {
+        // Compute final amount by subtracting  amount that was already billed.
+        if (!billedItems.iterator().hasNext() || billedUsage.compareTo(toBeBilledUsage) < 0) {
```

Comment:
Same remark regarding `!billedItems.iterator().hasNext()`.

Also, maybe `billedUsage.compareTo(toBeBilledUsage) < 0` should be extracted around these methods, since it's the same in both implementations:

```
if (billedUsage.compareTo(toBeBilledUsage) < 0) {
  populateResults(...);
}
```

Reply:
I think it's more tricky than it looks, moving that logic into `ContiguousIntervalUsageInArrear` would require reintroducing a bunch of `if mode = CONSUMABLE`, ...

Reply:
I don't think so, since it's the same logic for both modes.